jormungandr/src/explorer/mod.rs

```diff
@@ -225,21 +225,17 @@ impl ExplorerDB {
     /// chain length is greater than the current.
     /// This doesn't perform any validation on the given block and the previous state, it
     /// is assumed that the Block is valid
-    fn apply_block(
-        &mut self,
-        block: Block,
-    ) -> impl Future<Item = multiverse::Ref<State>, Error = Error> {
+    async fn apply_block(&mut self, block: Block) -> Result<multiverse::Ref<State>> {
         let previous_block = block.header.block_parent_hash();
         let chain_length = block.header.chain_length();
         let block_id = block.header.hash();
         let multiverse = self.multiverse.clone();
         let current_tip = self.longest_chain_tip.clone();
         let discrimination = self.blockchain_config.discrimination.clone();
 
-        multiverse
-            .get_ref(previous_block)
-            .map_err(|_: Infallible| unreachable!())
-            .and_then(move |maybe_previous_state| match maybe_previous_state {
+        let maybe_previous_state = multiverse.get_ref(previous_block).await;
+        let (parent_ref, transactions, blocks, addresses, epochs, chain_lengths, stake_pools) =
+            match maybe_previous_state {
```

Comment:
This could be made more readable by destructuring the `Some(State { .. })` directly into variables directly, without the monster tuple. The match expression could be transformed into `maybe_previous_state.ok_or_else(|| Error::from(...))?` to eliminate the explicit `None` arm.

Reply:
done